osm2streets/src/geometry/general_case.rs

```diff
@@ -43,6 +51,29 @@ pub fn trim_to_corners(
                 pt = one.pl.last_pt();
             }
 
+            // A special case happens when two roads share both endpoints, pointing in a loop. We
+            // may find a hit point on the wrong end, because there's no actual hit on the end
+            // we're currently looking.
+            let road1 = &roads[&one.road];
+            let road2 = &roads[&two.road];
+            if CommonEndpoint::new((road1.src_i, road1.dst_i), (road2.src_i, road2.dst_i))
+                == CommonEndpoint::Both
+            {
+                let other_i = if road1.src_i == results.intersection_id {
+                    road1.dst_i
+                } else {
+                    road1.src_i
+                };
+
+                let dist_to_this_intersection =
+                    intersection_endpoints[&results.intersection_id].dist_to(pt);
+                let dist_to_other_intersection = intersection_endpoints[&other_i].dist_to(pt);
+                if dist_to_other_intersection < dist_to_this_intersection {
```

Comment:
This is the best idea I've got for figuring out if we found a hit on the "correct" side. It might be brittle near super tiny roads or something. (But none of our other tests are affected at all by this change, so fine for now.) Any better ideas?

Reply:
After thinking "just check for intersections at the correct end, lol" and writing my comment, i was thinking about it while walking to the train, and realised "correct end" is difficult to define for arbitrary points along the road! This is the beat i can think of too.

If this becomes a problem, Maybe we can try to detect the "no intersection because the roads are pretty much parallel" case explicitly first. First check if the projection (infinite extension at this end) of both roads cross, then check if each edge endpoint sits within that projection. (Something along those lines, its still fuzzy in my head).